sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/CBDE/PreservingEncodingFallback.cs

```diff
@@ -0,0 +1,71 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2019 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Text;
+
+namespace SonarAnalyzer
+{
+    internal class PreservingEncodingBuffer : EncoderFallbackBuffer
+    {
+        public override int Remaining => buffer.Length - currentChar;
+
+        public override bool Fallback(char charUnknown, int index)
+        {
+            buffer = String.Format(".{0:X}", (int)charUnknown);
+            currentChar = 0;
+            return true;
+        }
+
+        public override bool Fallback(char charUnknownHigh, char charUnknownLow, int index)
+        {
+            buffer = String.Format(".{0:X}{1:X}", (int)charUnknownHigh, (int)charUnknownLow);
+            currentChar = 0;
+            return true;
+        }
+
+        public override char GetNextChar()
+        {
+            return currentChar < buffer.Length ? buffer[currentChar++] : '\u0000';
+        }
+
+        public override bool MovePrevious()
+        {
+            return false;
+        }
+        private string buffer;
+        private int currentChar;
+    }
+
+    /// <summary>
+    /// C# source code can contain any character, but MLIR only handle 8-bits chars. We must therefore encode C# names
+    /// so that two different strings in C# always result in two different strings in the generated code (by default, all
+    /// unknown characters would be translated to the same one)
+    /// </summary>
+    internal class PreservingEncodingFallback : EncoderFallback
```

Comment:
please move it in a separate file, we don't keep different types in a single file in our codebase and this is confusing